custom_components/google_home/api.py

```diff
@@ -236,73 +237,108 @@ async def delete_alarm_or_timer(
+                        response,
+                    )
+            else:
+                _LOGGER.error(
+                    (
+                        "Failed to get a confirmation that the %s"
+                        "was deleted for device %s. "
+                        "Received = %s"
+                    ),
+                    item_type,
+                    device.name,
+                    response,
+                )
+
+    async def reboot_google_device(self, device: GoogleHomeDevice) -> None:
+        """Reboots a Google Home device if it supports this."""
+
+        # "now" means reboot and "fdr" means factory reset (Not implemented).
+        data = {"params": "now"}
+
+        _LOGGER.debug(
+            "Trying to reboot Google Home device %s",
+            device.name,
+        )
+
+        response = await self.post(
+            endpoint=API_ENDPOINT_REBOOT, data=json.dumps(data), device=device
+        )
+
+        if response:
+            # It will return true even if the device does not support rebooting.
+            _LOGGER.info(
+                "Successfully asked %s to reboot.",
+                device.name,
+            )
+
+    async def post(
+        self, endpoint: str, data: str, device: GoogleHomeDevice
+    ) -> Optional[Dict[str, str]]:
+        """Shared post request"""
+
         if device.ip_address is None:
-            _LOGGER.warning("Device %s doesn't have IP address!", device.name)
-            return
+            _LOGGER.warning("Device %s doesn't have an IP address!", device.name)
+            return None
 
         if device.auth_token is None:
-            _LOGGER.warning("Device %s doesn't have a auth token!", device.name)
-            return
-
-        url = self.create_url(device.ip_address, PORT, API_ENDPOINT_DELETE)
-
-        # We need to remove charset=UTF-8 or else it will return a 400 Bad Request.
-        # I think this is because of the character "/" in the id string.
-        HEADERS.update(
-            {
-                HEADER_CAST_LOCAL_AUTH: device.auth_token,
-                HEADER_CONTENT_TYPE: "application/json",
-            }
-        )
+            _LOGGER.warning("Device %s doesn't have an auth token!", device.name)
+            return None
 
-        data = {"ids": [item_to_delete]}
+        url = self.create_url(device.ip_address, PORT, endpoint)
 
-        item_type = item_to_delete.split("/")[0]
+        HEADERS[HEADER_CAST_LOCAL_AUTH] = device.auth_token
 
         _LOGGER.debug(
-            "Deleting %s from Google Home device %s - %s - Raw data: %s",
-            item_type,
+            "Requesting endpoint %s for Google Home device %s - %s",
+            endpoint,
             device.name,
             url,
-            data,
         )
 
+        resp = None
+
         try:
             async with self._session.post(
-                url, json=data, headers=HEADERS, timeout=TIMEOUT
+                url, data=data, headers=HEADERS, timeout=TIMEOUT
```

Comment:
Why did you change `json` to `data`? Is it also works?

Reply:
Documentation says that it only adds "content-type: application/json", but I think setting headers explicitly redefines this anyway so it should be fine.

Reply:
`json` also converts it to JSON as far as i understand.

It is because of strict typing. If we use `json` we need to pass a known dict and we don't know this. This way we only need to define it to a `str`.

But maybe i am wrong about this. 🤔

Reply:
I see. Yeah, that's not easy to construct proper type for json and mypy still doesn't support recursive types which are needed for the correct type.

Reply:
This is a common problem i am running into with different libraries unfortunately. 😢 But by finding a fix I only get better at understanding python. 🚀 Also many places they use `Any` instead of explicitly stating the correct type.